### ISSUE TITLE
Files with Non-UTF8 characters are simply ignored in `ls`

### ISSUE THREAD

**Issue by theAkito** [nushell/nushell#11100]
### Describe the bug

Files with Non-UTF8 characters are simply ignored in `ls`.

### How to reproduce

1. Create a file named `''$'\001\b\327''@'$'\310\320\f''@8'`. Spelling and quoting is set from the point of view how Bash views this file's name.
2. `ls` with Nu.
3. See only normal files. This file is not displayed, at all.
4. `ls` with Bash.
5. File is shown.

### Expected behavior

Can view this file in some form or the other. Not even necessary to display all characters properly, I just want to view the file name, according to how Nu sees it, so I can `rm` it.

### Configuration

| key                | value                                                                     |
| ------------------ | ------------------------------------------------------------------------- |
| version            | 0.87.0                                                                    |
| branch             |                                                                           |
| commit_hash        | 77a1c3c7b2f3a110d48bcb792968e6b0d85d4bb7                                  |
| build_os           | linux-x86_64                                                              |
| build_target       | x86_64-unknown-linux-gnu                                                  |
| rust_version       | rustc 1.71.1 (eb26296b5 2023-08-03)                                       |
| rust_channel       | 1.71.1-x86_64-unknown-linux-gnu                                           |
| cargo_version      | cargo 1.71.1 (7f1d04c00 2023-07-29)                                       |
| build_time         | 2023-11-14 20:18:44 +00:00                                                |
| build_rust_channel | release                                                                   |
| allocator          | mimalloc                                                                  |
| features           | dataframe, default, extra, sqlite, static-link-openssl, trash, which, zip |
| installed_plugins  |                                                                           |

### Additional context

I accidentally created this file, saw it in VS Code.
Using Nu to list the file did not work out, at all. Then I `stat` the file, switched to Bash, rann `ls` there and voilà, there is that file. Removed it inside Bash.
Wasn't able to do it with Nu, because I could not even list, i.e. "see", it, which also means, I wasn't able to delete it.

**Comment by fdncred** [comment]
It's going to be hard to reproduce this without having some horked file like this.

**Comment by theAkito** [comment]
> It's going to be hard to reproduce this without having some horked file like this.

Isn't the bash `ls` representation of it enough?

**Comment by theAkito** [comment]
@fdncred

Pinging, in case you didn't see the edits to my previous comment, yet.

**Comment by fdncred** [comment]
ya, I saw it thanks. I was thinking that I may be able to create it with `touch $"(char ...)"`. I downloaded your zip and got this. I had a hard time unzipping it. unzip kept saying "Illegal byte sequence".

![image](https://github.com/nushell/nushell/assets/343840/54b03e57-43ff-4772-9503-d420aa3b286d)

**Comment by theAkito** [comment]
> I downloaded your zip and got this. I had a hard time unzipping it. unzip kept saying "Illegal byte sequence".

Apparently, it wasn't properly unzipped, because I could not see anything on my machine with the release version of 0.87.0.

I could try to encode the file or, if nothing seems to work, I can tell you how to reproduce the file yourself, but then you would need to install [Docker](https://www.docker.com/) & [Scuba](https://github.com/JonathonReinhart/scuba) on your machine.

I used Bash & `base64` to encode the file name. Maybe this helps?

Filename as Base64
```
AQjXQMjQDEA4Cg==
```

If that does not help, you can install the aforementioned software & run `scuba` in a directory, which features the following `.scuba.yml`.

```yaml
## https://scuba.readthedocs.io/en/latest/configuration.html
image: node:10-alpine3.11

hooks:
  root:
    script:
      - 'echo "HOOK: This runs before we switch users"'
      - id
  user: 'echo "HOOK: After switching users, uid=$(id -u) gid=$(id -g)"'

entrypoint: "/bin/sh"

environment:
  - "KEY=VALUE"
```

**Comment by fdncred** [comment]
It may be a problem/feature of macos, that it doesn't allow crazy filenames like this? I'm not sure. I'm also not using 0.87.0. I always use the latest main branch.

Since nushell is written in rust, there may be protections again invalid strings?
![image](https://github.com/nushell/nushell/assets/343840/73bcbf62-ec6a-410d-a0b2-a3f5def76817)

**Comment by theAkito** [comment]
> It may be a problem/feature of macos, that it doesn't allow crazy filenames like this? I'm not sure. I'm also not using 0.87.0. I always use the latest main branch.
> 
> ![image](https://user-images.githubusercontent.com/343840/284273633-73bcbf62-ec6a-410d-a0b2-a3f5def76817.png)

###### Bash
```bash
echo AQjXQMjQDEA4Cg== | base64 -d
�@��
    @8
```

###### Nu
```nu
"AQjXQMjQDEA4Cg==" | decode base64
Error:   × base64 payload isn't a valid utf-8 sequence
   ╭─[entry #1:1:1]
 1 │ "AQjXQMjQDEA4Cg==" | decode base64
   · ─────────┬────────
   ·          ╰── invalid utf-8 sequence of 1 bytes from index 2
   ╰────
  help: consider using the `--binary` flag

```

The latter seems to confirm, that Nu just cannot handle these characters, at all. It's strictly UTF-8 focused, disallowing anything else.
At the same time, Bash kind of handles it enough, so the user can at least deal with it.

> Since nushell is written in rust, there may be protections again[st] invalid strings?

Are you sure, there isn't a place in the core code, which defines some global to be strictly "UTF-8" or something?

Otherwise, if this is truly a Rust core inherent issue, I am very sure, this is not the first time someone encountered this issue.

**Comment by theAkito** [comment]
> So, in order to handle non-utf-8 strings, nushell would have to use something other than Rust's String/&str and nushell's `Value::String` is a rust `String`. You can see that here.

Okay. Nu's `ls` just ignores the file, in my case. I can neither notice it (without VS Code, I wouldn't even have noticed it's there), nor delete it (would require the displayed name in Nu's `ls` to work when piping it to Nu's `rm`).

Would it make sense to keep the default behaviour as it is & maybe add a flag to `ls` or some step deeper in the core code, where you can generally flag a string handling command to handle such weird scenarios, if needed?

Since this is a rare issue, it obviously does not have to be supported very well, but on the other hand, being completely "blind" about it in Nushell is not a solution, either. Like, if a user only uses Nu all the time, the data could be replicated & fill the folder with such files & the user wouldn't even notice, due to Nushell's ignorance.
If the file would be bigger, it could even lead to disk space issues, etc.

So, while this is a rare issue, if it ever appears, it may cause unforeseen circumstances.

**Comment by fdncred** [comment]
I agree that there should be a warning/error when doing `ls` that you have a non-utf-8 compliant file in your directory. The question is how to create that file for testing/debugging so someone can test and make changes to `ls` in order to display such a warning/error.

**Comment by theAkito** [comment]
> The question is how to create that file for testing/debugging so someone can test and make changes to `ls` in order to display such a warning/error.

Did you try the reproducible way I portrayed in the following comment?

https://github.com/nushell/nushell/issues/11100#issuecomment-1818983652

**Comment by fdncred** [comment]
> Did you try the reproducible way I portrayed in the following comment?

Nope. I'm not motivated enough to install docker or scuba.

**Comment by theAkito** [comment]
> > Did you try the reproducible way I portrayed in the following comment?
> 
> Nope. I'm not motivated enough to install docker or scuba.

Maybe someone else from the team has already half of it installed? 😉

**Comment by DonSheddow** [comment]
I could reproduce this using `touch "$(echo -ne "\xff\xff")"` in bash. Then `ls` in bash shows `''$'\377\377'`, while nu gives `warning: get non-utf8 filename "/tmp/test/\xFF\xFF", ignored.`

Note that Rust uses [OsString](https://doc.rust-lang.org/std/ffi/struct.OsString.html) to represent paths, which can handle invalid UTF-8.

For comparison, `os.listdir(".")` in Python returns `['\udcff\udcff']`, ie. it represents each invalid byte as a codepoint in the unicode surrogate block. When you want to open `'\udcff\udcff'` it just removes the prefix `\udc`, and this seems to be non-ambiguous because a surrogate codepoint does not have a valid UTF-8 representation. So this idea could be an option for nu if we don't want to introduce a separate datatype for paths.

**Comment by KAAtheWiseGit** [comment]
To add to this, once you get non-utf8 files somewhere, it's impossible to delete them in Nushell.  I had to fall back to Ash in order to execute a `rm *.docx` command.

**Comment by YizhePKU** [comment]
Non-utf8 filenames are pretty rare, but not negligible. Since one of the main responsibilities of a shell is to provide an interface to the underlying filesystem, I would argue that Nushell should handle non-utf8 filenames correctly.

**Comment by dxlr8r** [comment]
I have a similar issue, but with regular ASCII characters like tab and newline. They are recorded into ls's name row, but they cannot be used by other functions. Not the same issue with glob. From bash:

```
touch "$(printf 'nu\thates\nme')"
```

Then from NuShell:

```
> ls nu*
╭───┬─────────────────────┬──────┬──────┬──────────────╮
│ # │        name         │ type │ size │   modified   │
├───┼─────────────────────┼──────┼──────┼──────────────┤
│ 0 │ nu\u{9}hates\u{a}me │ file │  0 B │ a minute ago │
╰───┴─────────────────────┴──────┴──────┴──────────────╯
> ls nu* | get name | path exists
╭───┬───────╮
│ 0 │ false │
╰───┴───────╯
```
